### PR TITLE
feat(new transform): New `explode` transform

### PR DESCRIPTION
**Pull request by sech9446**
Ref: #6459 

Signed-off-by: Eunchul Song <eunchulsong9@gmail.com>

### CODE REVIEW

File: src/internal_events/explode.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#[derive(Debug)]
+pub struct ExplodeFieldIsNotArray<'a> {
+    pub field: &'a str,
+    pub kind: &'a str,
+}
+
+impl<'a> InternalEvent for ExplodeFieldIsNotArray<'a> {
+    fn emit_logs(&self) {
+        warn!(
+            message = "Cannot explode non array kind field.",
```
Comment (jszwedko) [inline src/internal_events/explode.rs:43]:
```suggestion
            message = "Cannot explode non-array field.",
```

File: src/transforms/explode.rs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        let target_field = if let Some(rename_as) = &self.rename_as {
+            rename_as
+        } else {
+            &self.field
+        };
+        if let Some(value) = event.as_mut_log().remove_prune(&self.field, true) {
+            if let Value::Array(array) = value {
+                if event
+                    .as_mut_log()
+                    .insert(&target_field, Value::Null)
```
Comment (jszwedko) [inline src/transforms/explode.rs:83]: I think you could just use `.get()` here to avoid the unnecessary insert operation.

**Review by jszwedko — COMMENTED** [review]
This is great work, thanks @sech9446 ! Thank you for adding all of the tests.

I left a few small comments. I'm also wondering if we might be able to expand the functionality of this transform to handle arrays of objects differently.

I'm imagining a case where a user might have input like:

```
{
  "events": [
    {
      "message": "event 1"
    },
    {
      "message": "event 2"
    }
  ]
}
```

And they want to output:

```
{
  "message": "event 1"
}
```
and
```
{
  "message": "event 2"
}
```

I think this could be handled by checking the type of the array value as we are iterating and, for objects, instead of inserting as a child field, merge with the original event.

I could see this being a configurable option so that users could still have it as a child field when desired, which is how you currently have it.

This will also need docs, but I can help with that once it is ready to be merged.

**Comment by sech9446** [comment]
Thank you for the detailed review!

Your suggestion make sense. I am going to add two new configurable options `drop_invalid` and `flatten`, with default value `false` of both.

**Comment by jszwedko** [comment]
@sech9446 thanks for your work on this! We used it as inspiration for the `unnest` function in #7404 which, combined with #7267 will let users do this transformation just in `remap`. I'll close this out given the other approach, but let us know if that approach doesn't work for you.

**Pull request closed without merging**